packages/design-system/src/components/Button/Button.tsx

```diff
@@ -42,35 +53,25 @@ type CommonButtonProps = {
   variation?: ButtonVariation;
 };
 
-type OmitProps = 'children' | 'className' | 'onClick' | 'ref' | 'size' | 'type' | 'href';
-type PropsOf<T extends ButtonComponentType> = Omit<React.ComponentPropsWithRef<T>, OmitProps>;
-
-export type CustomButtonComponentType = React.ComponentType<any> | React.FC;
-export type ButtonComponentType = React.ElementType<any> | CustomButtonComponentType;
-export type NonAnchorButtonComponentType =
-  | CustomButtonComponentType
-  | React.ElementType<Exclude<string, 'a' | 'button'>>;
-
-type LinkTypeButtonProps = {
-  component?: 'a';
-  href: string;
-};
-type DefaultButtonTypeButtonProps = {
-  component?: 'button';
-  href?: undefined | null;
-};
-type OtherTypeButtonProps<T extends NonAnchorButtonComponentType> = {
-  component: T;
-  href?: undefined | null;
-};
-
-export type ButtonProps<T extends ButtonComponentType> = PropsOf<T> &
-  CommonButtonProps &
-  (
-    | LinkTypeButtonProps
-    | DefaultButtonTypeButtonProps
-    | (T extends NonAnchorButtonComponentType ? OtherTypeButtonProps<T> : never)
-  );
+// Collect all the additional properties that one could supply to a button component
+// that will be passed down to whatever element or component is being used. This is
+// generally permissive in order to keep the typing simple at the expense of being
+// more accurate. `OtherProps` is generic so that we can extract any props from a
+// custom component that is being passed in. I'm trying to keep most of the complexity
+// in this section and leave the `ButtonProps` definition simpler. - PW
+//
+// Extend is a utility type that works like `Object.assign` where properties defined
+// on the latter type `N` override properties defined on the former type `M`
+type Extend<M, N> = Omit<M, Extract<keyof M, keyof N>> & N;
+type OtherProps<T extends ButtonComponentType> = Omit<
+  // Get all possible HTML attributes and override with any more specific props from
+  // the possibly custom component type `T`
+  Extend<React.HTMLAttributes<HTMLElement>, React.ComponentPropsWithRef<T>>,
+  // And omit any properties that we're defining on our own `CommonButtonProps`
+  keyof CommonButtonProps<T>
+>;
+
+export type ButtonProps<T extends ButtonComponentType> = CommonButtonProps<T> & OtherProps<T>;
 
 export const Button = <T extends ButtonComponentType>(props: ButtonProps<T>) => {
```

Comment:
Could this can be simplified to:

`export const Button = (props: ButtonProps<ButtonComponentType>) => {`

Reply:
The hope here is to get specific props for the `component` component, like in https://github.com/CMSgov/design-system/blob/pwolfert/fix-button-types-5/packages/design-system/src/components/Button/Button.test.tsx#L53-L60

Reply:
I see gotcha!

Reply:
I also just double-checked to make sure the generic type on `CommonPropTypes` was necessary in order to infer the actual type of `component`, and it was. If I remove it, it allows the `Link` unit test to omit the `to` prop.